### PR TITLE
Added Propagation of ManagedNodeGroup Tags to their corresponding AutoScalingGroups

### PR DESCRIPTION
**Pull request by SlevinWasAlreadyTaken**
#### Resume

I have added Propagation of ManagedNodeGroup (MNG) Tags to their AutoScalingGroups (ASG). This should resolve the issue #1571 

- The changes ensure that AutoScalingGroups Tags are the same as their corresponding ManagedNodeGroup after it has been automatically created by AWS.
- All tags are copied from the ManagedNodeGroup to the AutoScalingGroup. If the tags already exists, it is overridden.
- The propagation is enabled by default (as it is for Unmanaged NodeGroup) and it can be disabled using the DisableASGTagPropagation boolean in the configuration.

#### How it has been done

I tried to find a good place to do it, and added it as an additional task on managed group creation. I switched the Parallel task boolean to false since it requires to be Sequential: we can't propagate the MNG tags to the ASGs without having the MNG created (since the ASGs are created automatically after the MNG creation).

ℹ️ This is my first contribution, thanks for your understanding if I've missed something =)

### Checklist
- [x] Added tests that cover your change (if possible)
    ~~- ⚠️ As it is my first contribution, I would prefer a first review before of course, implementing them~~
- [x] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes

### CODE REVIEW

**Comment by afirth** [comment]
Nice one @SlevinWasAlreadyTaken !
I agree, it makes sense to hold off on automatically generating these tags for taints and labels until #4853 is resolved

**Comment by afirth** [comment]
@aclevername @Himangini will be interested I guess :100:

**Comment by SlevinWasAlreadyTaken** [comment]
I have rebased this PR considering [this fresh changes](https://github.com/weaveworks/eksctl/commit/4435d5636ad94b7b485700f8e35787d3dd2ba53e) in the way we configure the propagation of tags.

File: pkg/apis/eksctl.io/v1alpha5/assets/schema.json
Comment (Skarlso) [inline pkg/apis/eksctl.io/v1alpha5/assets/schema.json:1322]: Why is this deleted? 🤔

File: pkg/cfn/manager/create_tasks.go
Comment (Skarlso) [inline pkg/cfn/manager/create_tasks.go:95]: Why is this changed?

File: pkg/cfn/manager/nodegroup.go
Comment (Skarlso) [inline pkg/cfn/manager/nodegroup.go:93]: Please add a check if the number of tags on the ASG would go over the limit of 50 and throw an error in that case.

File: pkg/cfn/manager/nodegroup.go
Comment (Skarlso) [inline pkg/cfn/manager/nodegroup.go:112]: as with other tasks, this whole thing should be in a go routine with a defer close on the err channel.

Comment (Skarlso) [reply pkg/cfn/manager/nodegroup.go:112]: Try to take a look at `createClusterTask`.

Comment (Skarlso) [reply pkg/cfn/manager/nodegroup.go:112]: Alternatively, we do tasks now differently. :D But I'm not going to force that on this contribution, since it's not that trivial to implement out of the blue. :) But if you want to, things are under `pkg/actions` now-a-days.

Take a look at pkg/actions/karpenter for tasks and how they are initialised. If you wish. :)

File: pkg/cfn/manager/tasks_test.go
Comment (Skarlso) [inline pkg/cfn/manager/tasks_test.go:115]: Why change this? This should be a new test, right? The original `makeManagedNodeGroups` task didn't disappear, did it?

File: userdocs/src/usage/eks-managed-nodes.md
Comment (Skarlso) [inline userdocs/src/usage/eks-managed-nodes.md:249]: I wouldn't completely delete this sentence as they still propagate. Just change the last entry to say, `and to the Autoscaling Group.`

**Review by Skarlso — CHANGES_REQUESTED** [review]
Couple changes required, otherwise this is on a good track! :) Thank you!

File: userdocs/src/usage/autoscaling.md
Comment (afirth) [inline userdocs/src/usage/autoscaling.md:47]:
```suggestion
For unmanaged and managed nodegroups, this is done by `eksctl` automatically if `propagateASGTags` is set to `true` like this:
```

File: userdocs/src/usage/eks-managed-nodes.md
Comment (afirth) [reply userdocs/src/usage/eks-managed-nodes.md:249]: I think this is deleted because it is in the "unsupported options" aka "limitations" section - which with this PR it no longer is

File: pkg/apis/eksctl.io/v1alpha5/assets/schema.json
Comment (SlevinWasAlreadyTaken) [reply pkg/apis/eksctl.io/v1alpha5/assets/schema.json:1322]: I thought, since it was deprecated, that the doc could be removed to avoid people using it. I realise now it was not done for the non-managed nodegroup. I did put it back then to keep consistent.

File: pkg/cfn/manager/create_tasks.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/create_tasks.go:95]: The task cannot be done in parallel of the managed nodegroup creation since the auto-scaling group is created by AWS after the managed nodegroup has been created.

File: userdocs/src/usage/eks-managed-nodes.md
Comment (SlevinWasAlreadyTaken) [reply userdocs/src/usage/eks-managed-nodes.md:249]: @afirth answered to this https://github.com/weaveworks/eksctl/pull/5002#discussion_r841743039

There not no longer limitations about it: cf "The unsupported options are noted below."

Do you still want me to re-add it?

File: pkg/cfn/manager/nodegroup.go
Comment (afirth) [reply pkg/cfn/manager/nodegroup.go:93]: https://docs.aws.amazon.com/autoscaling/ec2/userguide/autoscaling-tagging.html#tag_restrictions @SlevinWasAlreadyTaken

Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/nodegroup.go:90]: Done in 0099c487ccebe4d8141505c8db17fcaa3f158ff0
Checked both the 50 limit and the 25 (creating tags by chunk of 25 tags if necessary)

File: pkg/cfn/manager/nodegroup.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/nodegroup.go]: Done in 0099c487ccebe4d8141505c8db17fcaa3f158ff0
Tried to do it the "old way" for now :)

File: pkg/cfn/manager/tasks_test.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/tasks_test.go]: This is true, re-added in 34516fb53077db5587bbcaff86a6ed5566ea1676

File: userdocs/src/usage/autoscaling.md
Comment (SlevinWasAlreadyTaken) [reply userdocs/src/usage/autoscaling.md]: fixed in 69390171dcd08ca0f96a6da8ebc8092a235e0cf8

File: userdocs/src/usage/eks-managed-nodes.md
Comment (Skarlso) [reply userdocs/src/usage/eks-managed-nodes.md:249]: Nah, it's fine. :)

**Comment by Skarlso** [comment]
@SlevinWasAlreadyTaken hi. You need to resolve the merge so I can run the tests. :) thanks!

**Comment by SlevinWasAlreadyTaken** [comment]
> @SlevinWasAlreadyTaken hi. You need to resolve the merge so I can run the tests. :) thanks!

I just rebased!

File: pkg/cfn/manager/create_tasks.go
Comment (Skarlso) [reply pkg/cfn/manager/create_tasks.go:95]: Hm, I have to take a look at this, but this means that if you have multiple nodegroups this will significantly slow down eksctl since all of them will be created in parallel. I think the ASG creation is part of this task so that should be fine. Did you test it by any chance leaving it like that?

Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/create_tasks.go:95]: Going to perform some testing.

**Comment by SlevinWasAlreadyTaken** [comment]
@Skarlso I realise after my rebase that the tests are broken, but it seems to be related to code I didn't touch. Any clue about it? There was only on line conflicting and it was about a go import 🤔

File: pkg/apis/eksctl.io/v1alpha5/assets/schema.json
Comment (Skarlso) [reply pkg/apis/eksctl.io/v1alpha5/assets/schema.json:1323]: Yeah, we leave things around for a little while before deleting. :) It's a deprecation tactic.

**Comment by Skarlso** [comment]
@SlevinWasAlreadyTaken There is something wrong with the merge? 

[](https://github.com/weaveworks/eksctl/runs/5833661917?check_suite_focus=true)Check failure on line 284 in pkg/cfn/manager/api.go

GitHub Actions
/ Unit tests
pkg/cfn/manager/api.go#L284
not enough arguments in call to c.asgAPI.CreateOrUpdateTags

File: pkg/cfn/manager/create_tasks.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/create_tasks.go:95]: Without sequential tasks, it fails since the node group does not exist yet while we try to create the ASG.
```bash
2022-04-05 15:04:11 [✖]  couldn't get managed nodegroup details for nodegroup "test-m-4xlarge-worker-c-v21-1test": ResourceNotFoundException: No node group found for name: test-m-4xlarge-worker-c-v21-1test.
{
  RespMetadata: {
    StatusCode: 404,
    RequestID: "ca88f3b0-2c12-46e3-b577-aa7db73daef2"
  },
  ClusterName: "test",
  Message_: "No node group found for name: test-m-4xlarge-worker-c-v21-1test.",
  NodegroupName: "test-m-4xlarge-worker-c-v21-1test"
}
```

Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/create_tasks.go:95]: Creating two nodegroup with the propagation, I see I have 4 sequential tasks.

I was expecting to have 2 parallels tasks corresponding to the 2 nodegroups, having inside 2 sequential tasks (ng creation + tags propagation). If you have any idea about how to do that, thanks to let me know.

For now, I have mitigated the slowness, setting back Parallel to `true` if no managednodegroup has at least one propagateASGTags set to true.

**Comment by SlevinWasAlreadyTaken** [comment]
@Skarlso Indeed.
The switch to aws-sdk-go-v2 went wrong with the rebase.
Fixed in d1255db87e1094a500a9806fd523e32879e662ec + 057ea8bda39f00c827d938c5c56c079cbfd4adeb

File: pkg/cfn/manager/create_tasks.go
Comment (Skarlso) [reply pkg/cfn/manager/create_tasks.go]: Uh, I have no idea. :D the task system is pretty complex. I'll try to unravel it today.

File: pkg/cfn/manager/api.go
Comment (Skarlso) [inline pkg/cfn/manager/api.go:312]: If there are duplicates, we should error right away like here: https://github.com/weaveworks/eksctl/blob/main/pkg/cfn/builder/nodegroup.go#L302

Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/api.go:312]: As the ASG is automatically created, it has some usual tags on it such as the cluster-name it runs on for example. This is normal and we don't want to throw an error about it.

File: pkg/cfn/manager/api.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/api.go:242]: Indeed. I didn't add unit tests yet but I'll do it today. I was waiting for a more global review on this contribution :) Keep you posted 👍

Comment (Skarlso) [reply pkg/cfn/manager/api.go:242]: Ah okay. Otherwise, it looks fine now, I think. :) I'll let the other members also review this. :) Thanks for your hard work!

File: pkg/cfn/manager/api.go
Comment (Skarlso) [reply pkg/cfn/manager/api.go:312]: Ah gotcha. That's interesting. I didn't think of that.

File: pkg/cfn/manager/api.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/api.go:242]: @Skarlso just added the unit tests for you information.

File: pkg/cfn/manager/api_test.go
Comment (Skarlso) [inline pkg/cfn/manager/api_test.go:28]: I was about to say extract common steps into a `BeforeEach` but seeing that nothing else is doing that in this test, I'd say meh. :D

Comment (Skarlso) [reply pkg/cfn/manager/api_test.go:28]: Ah, actually, yes! Recent change -> `GetClusterStackIfExists`. Can you please do the same? :))) For a better structure. The test looks good to me.

**Review by Skarlso — APPROVED** [review]
LGTM after conflict and comment resolution.

It needs another set of eyes from the Team. :)

File: pkg/cfn/manager/api_test.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/api_test.go]: @Skarlso Done in 8e5106fb5cba2f3df808523db670552af1662418
I have used `BeforeEach` to init the variables that could be take advantage of it =)

Comment (Skarlso) [reply pkg/cfn/manager/api_test.go]: Thanks!

File: pkg/cfn/manager/api.go
Comment (Himangini) [inline pkg/cfn/manager/api.go:10]:
```suggestion
	asTypes "github.com/aws/aws-sdk-go-v2/service/autoscaling/types"
```

small nit

File: pkg/cfn/manager/api_test.go
Comment (Himangini) [inline pkg/cfn/manager/api_test.go:8]:
```suggestion
	asTypes "github.com/aws/aws-sdk-go-v2/service/autoscaling/types"
```

small nit

**Review by Himangini — APPROVED** [review]
LGTM 👍🏻

File: pkg/cfn/manager/api_test.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/api_test.go]: done in 83dfd587db2ddad0a41260db93b2475ce3248cb5

File: pkg/cfn/manager/api.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/api.go]: done in 83dfd587db2ddad0a41260db93b2475ce3248cb5

**Review by nikimanoledaki — CHANGES_REQUESTED** [review]
Well done on this PR! 🎉 I left some comments and suggestions for a few possible improvements. Other than that, this PR is looking great, thank you for your contribution! 😄 
Note: the request for change is for the top-level file `dev-managed.yaml`.

File: dev-managed.yaml
Comment (SlevinWasAlreadyTaken) [reply dev-managed.yaml]: Indeed, this is the file for testing purpose. I removed it.

File: pkg/cfn/manager/api.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/api.go]: done in e7877a29358e7c3ec69a7dba0192c0346b5a3df9

File: pkg/cfn/manager/api.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/api.go]: done in e7877a29358e7c3ec69a7dba0192c0346b5a3df9

File: pkg/cfn/manager/nodegroup.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/nodegroup.go]: Indeed ! done in 4e2645b666fd81eea33161b2cdcc06280eb0b709

File: pkg/cfn/manager/api.go
Comment (SlevinWasAlreadyTaken) [reply pkg/cfn/manager/api.go]: Done in 4e2645b666fd81eea33161b2cdcc06280eb0b709

**Review by nikimanoledaki — APPROVED** [review]
Great! Thank you for addressing the change requests. LGTM once the merge conflicts are addressed 😄 🎉

**Comment by SlevinWasAlreadyTaken** [comment]
@nikimanoledaki 👋 conflicts addressed :)

**Comment by Skarlso** [comment]
@SlevinWasAlreadyTaken thank you for your contribution and sticking to this long process. :) Well done! :)

**Pull request merged**
